src/networking.c

```diff
@@ -4775,6 +4895,18 @@ void ioThreadReadQueryFromClient(void *data) {
         c->io_parsed_cmd = NULL;
     }
 
+    /* Offload slot calculations to the I/O thread to reduce main-thread load. */
+    if (c->io_parsed_cmd && server.cluster_enabled) {
+        getKeysResult result;
+        initGetKeysResult(&result);
+        int numkeys = getKeysFromCommand(c->io_parsed_cmd, c->argv, c->argc, &result);
+        if (numkeys) {
+            robj *first_key = c->argv[result.keys[0].pos];
+            c->slot = calculateKeySlot(first_key->ptr);
+        }
```

Comment:
Is it safe to do this before the validation logic in `getNodeByQuery` operation? There could be multi slot keys which is invalid.

Reply:
This should be safe to do.

> In getNodeByQuery, we override this value.

Ideally we wouldn't recompute this unnecessarily, but I suppose it's not too expensive. Do we have a list of potential optimizations we're tracking once this is done?